File: lidar_camera_calibration/lidar_cam_ws/src/pcl_time_change/src/pcl_time_change.cpp
#include <ros/ros.h>
#include <sensor_msgs/PointCloud2.h>

ros::Publisher pub;
int time_threshold;

void callback(const sensor_msgs::PointCloud2ConstPtr& msg) {
  sensor_msgs::PointCloud2 new_msg = *msg;
  ROS_INFO("original pointcloud time: %d.%d", new_msg.header.stamp.sec, new_msg.header.stamp.nsec);
  if (new_msg.header.stamp.nsec < time_threshold) {
    new_msg.header.stamp.sec -= 1;
    new_msg.header.stamp.nsec += 1000000000 - time_threshold;  // add 0.6 seconds
  }else{
    new_msg.header.stamp.nsec -= time_threshold;  // subtract 0.4 seconds
  }

  pub.publish(new_msg);
}

int main(int argc, char** argv) {
  ros::init(argc, argv, "pointcloud_modifier");
  ros::NodeHandle nh;
  ros::NodeHandle nh_private("~");

  std::string input_topic, output_topic;
    nh_private.param<std::string>("input_topic", input_topic, "/input_pointcloud");
    nh_private.param<std::string>("output_topic", output_topic, "/output_pointcloud");
    nh_private.param<int>("time_threshold", time_threshold, 400000000);

  ROS_INFO("pointcloud_TIME_modifier node started");
  pub = nh.advertise<sensor_msgs::PointCloud2>(output_topic, 1);
  ros::Subscriber sub = nh.subscribe(input_topic, 1, callback);

  ros::spin();
  return 0;
}

File: format_conversion_git_ws/src/pcl_pkgs/pcl_time_change/script/pcl_time_change.py
#!/usr/bin/env python

import rospy
from sensor_msgs.msg import PointCloud2, PointField
from std_msgs.msg import Header
import sys

class PointCloudProcessor:
    def __init__(self):
        # Initialize the node
        rospy.init_node('point_cloud_processor', anonymous=True)

        # Set up the subscriber and publisher
        self.subscriber = rospy.Subscriber('/input_point_cloud', PointCloud2, self.callback)
        self.publisher = rospy.Publisher('/output_point_cloud', PointCloud2, queue_size=10)

    def callback(self, data):
        # Create a new PointCloud2 message
        new_point_cloud = PointCloud2()

        # Subtract 100 ms from the header timestamp
        new_time = data.header.stamp - rospy.Duration(0.1)

        # Update the new PointCloud2 message with the modified header and original data
        new_point_cloud.header = Header(data.header.seq, new_time, data.header.frame_id)
        new_point_cloud.height = data.height
        new_point_cloud.width = data.width
        new_point_cloud.fields = data.fields
        new_point_cloud.is_bigendian = data.is_bigendian
        new_point_cloud.point_step = data.point_step
        new_point_cloud.row_step = data.row_step
        new_point_cloud.data = data.data
        new_point_cloud.is_dense = data.is_dense

        # Publish the modified point cloud
        self.publisher.publish(new_point_cloud)

def main(args):
    try:
        pc_processor = PointCloudProcessor()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass

if _name_ == '_main_':
    main(sys.argv)